knnaas/worker/KNNIndex.cpp

```diff
@@ -0,0 +1,415 @@
+
+		    signal(SIGSEGV, handler);
+		    ScopedGILRelease release_gil = ScopedGILRelease();
+		    
+            int total_threads = (remainder == 0) ? num_threads : num_threads + 1;
+
+		    v_results.reserve(total_threads);
+		    v_threads.reserve(total_threads);
+
+            // TODO: Use threadpool instead. There is an overhead with thread creation and destruction.
+		    for (int i = 0; i < total_threads; i++) {
+			    v_results.push_back(std::vector<ResultRecord>());
+			    std::vector<ResultRecord>& res = v_results[i];
+			    end_index = (i < num_threads) ? begin_index + batch_size : begin_index + (index.size() - begin_index);
+			    v_threads.push_back(
+                                    std::thread(
+                                        &KNNIndex::calculate_distance_return_list_using_iterator, 
+                                            this, 
+                                            i, 
+                                            std::ref(input_vector), 
+                                            std::ref(input_tags), 
+                                            begin_index, 
+                                            end_index,  
+                                            std::ref(res), 
+                                            top_n
+                                            )
+                                   );
+			    begin_index += batch_size;
+            }
+            
+            std::for_each(v_threads.begin(), v_threads.end(), std::mem_fn(&std::thread::join));
+		    std::priority_queue<ResultRecord, std::vector<ResultRecord>, ResultRecord> top_k;
+            
+            for(int i = 0; i < v_results.size(); i++) {
+                for (int j = 0; j < v_results[i].size(); j++) {
+				    ResultRecord& res = v_results[i][j];
+				    top_k.push(res);
+
+				    if (top_k.size() > top_n) {
+					    top_k.pop(); // TODO : how do i remove the element from the back? top_back?
+				    }
+			    }
+		    }
+            
+		    std::vector<ResultRecord> res;
+		    int result_count = (top_k.size() < top_n) ? top_k.size() : top_n;
+
+		    while (res.size() < result_count) {
+			    res.push_back(top_k.top());
+			    top_k.pop();
+		    }
+		    
+            return res;
+	}
+
+    void calculate_distance_return_list_using_iterator(
+                                                        int thread_id, 
+                                                        std::vector<float>& input_vector,
+                                                        std::vector<long>& input_tags,
+                                                        long begin_index,
+                                                        long end_index,
+                                                        std::vector<ResultRecord>& result,
+                                                        int top_n
+                                                       ) {
+            float dist;
```

Comment:
@akshay-sharma 
Overflow can occur with doubles as well, though less likely. 
Better to test for overflow and underflow conditions before returning.